sys/shell/commands/sc_gnrc_icmpv6_echo.c

```diff
@@ -0,0 +1,432 @@
+        .count = DEFAULT_COUNT,
+        .tmin = UINT_MAX,
+        .datalen = DEFAULT_DATALEN,
+        .timeout = DEFAULT_TIMEOUT_USEC,
+        .interval = DEFAULT_INTERVAL_USEC,
+        .id = DEFAULT_ID,
+        .pattern = DEFAULT_ID,
+    };
+    int res;
+
+    if ((res = _configure(argc, argv, &data)) != 0) {
+        return res;
+    }
+    gnrc_netreg_register(GNRC_NETTYPE_ICMPV6, &data.netreg);
+    _pinger(&data);
+    do {
+        msg_t msg;
+
+        msg_receive(&msg);
+        switch (msg.type) {
+            case GNRC_NETAPI_MSG_TYPE_RCV: {
+                _handle_reply(&data, msg.content.ptr);
+                gnrc_pktbuf_release(msg.content.ptr);
+                break;
+            }
+            case _SEND_NEXT_PING:
+                _pinger(&data);
+                break;
+            case _PING_FINISH:
+                goto finish;
+            default:
+                /* requeue wrong packets */
+                msg_send(&msg, sched_active_pid);
+                break;
+        }
+    } while (data.num_recv < data.count);
+finish:
+    xtimer_remove(&data.sched_timer);
+    res = _finish(&data);
+    gnrc_netreg_unregister(GNRC_NETTYPE_ICMPV6, &data.netreg);
+    for (unsigned i = 0;
+         i < cib_avail((cib_t *)&sched_active_thread->msg_queue);
+         i++) {
+        msg_t msg;
+
+        /* remove all remaining messages (likely caused by duplicates) */
+        if ((msg_try_receive(&msg) > 0) &&
+            (msg.type == GNRC_NETAPI_MSG_TYPE_RCV) &&
+            (((gnrc_pktsnip_t *)msg.content.ptr)->type == GNRC_NETTYPE_ICMPV6)) {
+            gnrc_pktbuf_release(msg.content.ptr);
+        }
+        else {
+            /* requeue other packets */
+            msg_send(&msg, sched_active_pid);
+        }
+    }
+    return res;
+}
+
+static void _usage(char *cmdname)
+{
+    printf("%s [-c <count>] [-h] [-i <ms interval>] [-s <packetsize>]\n",
+           cmdname);
+    puts("     [-t hoplimit] [-W <ms timeout>] <host>[%<interface>]");
+    puts("     count: number of pings (default: 3)");
+    puts("     ms interval: wait interval milliseconds between sending "
+              "(default: 1000)");
+    puts("     packetsize: number of bytes in echo payload; must be >= 4 to "
+              "measure round trip time (default: 4)");
+    puts("     hoplimit: Set the IP time to life/hoplimit "
+              "(default: interface config)");
+    puts("     ms timeout: Time to wait for a resonse in milliseconds "
+              "(default: 1000). he option affects only timeout in absence "
+              "of any responses, otherwise wait for two RTTs");
+}
+
+static int _configure(int argc, char **argv, _ping_data_t *data)
+{
+    char *cmdname = argv[0];
+    int res = 0;
+
+    /* jump over cmdname */
+    if ((--argc) == 0) {
+        /* only command name given */
+        res = 1;
+    }
+    ++argv;
+    /* parse command line arguments */
+    for (; argc > 0; argc--, argv++) {
```

Comment:
I find this simultaneous decrementing and incrementing kind of fragile. Why not use an explicit index?

Reply:
For `argc`/`argv` I find this usually ok, but let me try with implicit index.

Reply:
Ok, going for explicit index.